data/trie/patriciaMerkleTrie.go

```diff
@@ -585,50 +591,36 @@ func (tr *patriciaMerkleTrie) GetSerializedNodes(rootHash []byte, maxBuffToSend
 	return nodes, remainingSpace, nil
 }
 
-// GetAllLeaves iterates the trie and returns a map that contains all leafNodes information
-func (tr *patriciaMerkleTrie) GetAllLeaves() (map[string][]byte, error) {
-	tr.mutOperation.RLock()
-	defer tr.mutOperation.RUnlock()
-
-	if tr.root == nil {
-		return map[string][]byte{}, nil
-	}
-
-	leaves := make(map[string][]byte)
-	err := tr.root.getAllLeaves(leaves, []byte{}, tr.Database(), tr.marshalizer)
-	if err != nil {
-		return nil, err
-	}
-
-	return leaves, nil
-}
-
 // GetAllLeavesOnChannel adds all the trie leaves to the given channel
-func (tr *patriciaMerkleTrie) GetAllLeavesOnChannel() chan core.KeyValueHolder {
+func (tr *patriciaMerkleTrie) GetAllLeavesOnChannel(rootHash []byte) (chan core.KeyValueHolder, error) {
 	//TODO pass a context for cancellation purposes when needed
-	leavesChannel := make(chan core.KeyValueHolder)
+	leavesChannel := make(chan core.KeyValueHolder, 100)
 
 	tr.mutOperation.RLock()
-	if tr.root == nil {
+
+	newTrie, err := tr.recreate(rootHash)
+	if err != nil || check.IfNil(newTrie) || tr.root == nil {
 		tr.mutOperation.RUnlock()
 		close(leavesChannel)
-
-		return leavesChannel
+		return leavesChannel, err
```

Comment:
Done.

Reply:
Actually, will leave it like this. The reason is that doing a range over a nil channel results in a deadlock, but doing a range on a closed channel (as it is the case with this) will do nothing (as it should).

Reply:
Yeah, the idea we use in the whole project is that: if the returned error is not nil, do not bother with the returned pointer as it might be nil or holding a wrong value

Reply:
Refactored.